Migrate your MySQL data using ``mysqldump`` (#1345)
aiven/devportal: docs/products/mysql/howto/migrate-database-mysqldump.rst

> Migrate your MySQL data using ``mysqldump``

Reviewer: I would say the perfect title is "Backup and restore MySQL data using ``mysqldump``"

===========================================

Performing the backup of your MySQL data to another storage service is a good way to ensure access to your data in case a failure occurs. In this article, you can find out how to copy your Aiven for MySQL data to a file and backup to another Aiven for MySQL database using ``mysqldump`` `tool <https://dev.mysql.com/doc/refman/8.0/en/mysqldump.html>`__.

Prerequisites
-------------

* The ``mysqldump`` `tool <https://dev.mysql.com/doc/refman/8.0/en/mysqldump.html>`_ installed. Check out the `official MySQL <https://dev.mysql.com/doc/mysql-shell/8.0/en/mysql-shell-install.html>`_ documentation on how to install it.
  
* A MySQL database that we aim to copy our data from. We will refer to it as ``source-db``.
  
* A target MySQL database that we aim to dump our ``source-db`` data. We will refer to it as ``target-db``.

.. tip::

>     For the dumping process, we recommend you pick a plan size that is large enough to store your data, so you can limit the downtime during migration.

Reviewer: ```suggestion
    For the restore process, we recommend you pick a plan size that is large enough to store your data, so you can limit the downtime if you're performing a migration.
```

> To simplify the example, we'll create Aiven for MySQL databases for both, ``source-db`` and ``target-db``. You can create the databases by going to your running service for **Aiven for MySQL** > **Databases** > **Create a new database**.

Reviewer: ```suggestion
The below example uses Aiven for MySQL databases for both, ``source-db`` and ``target-db``. You can create the databases by going to your running service for **Aiven for MySQL** > **Databases** > **Create a new database**.
```

Backup the data
---------------

> First, we'll back up our ``source-db`` data to a file called ``mydb_backup.sql``. For that, you need to collect some information about your Aiven for MySQL ``source-db`` database. Go to your Aiven ``source-db`` service, from **Overview** page, find the following information:

Reviewer: ```suggestion
To backup the ``source-db`` data to a file called ``mydb_backup.sql`` you need to collect some information about your Aiven for MySQL ``source-db`` database. Go to your Aiven ``source-db`` service, from **Overview** page, find the following information:
```

.. list-table::
   :widths: 20 80
   :header-rows: 1

   * - Variable
     - Description
   * - ``SOURCE_DB_HOST``
     - **Host** name for the connection
   * - ``SOURCE_DB_USER``
     - **User** name for the connection
   * - ``SOURCE_DB_PORT``
     - Connection **Port** number
   * - ``SOURCE_DB_PASSWORD``
     - Connection **Password**
   * - ``DEFAULTDB``
     - Database that contains the ``source-db`` data

Use the following command to backup your Aiven for MySQL data to the ``mydb_backup.sql`` file:

.. code-block:: shell

    mysqldump \
    -p DEFAULTDB -P SOURCE_DB_PORT \
    -h SOURCE_DB_HOST --single-transaction \
    -u SOURCE_DB_USER --set-gtid-purged=OFF \
    --password > mydb_backup.sql

With this command, the password will be requested at the prompt; paste ``SOURCE_DB_PASSWORD`` to the terminal, then a file named ``mydb_backup.sql`` will be created with your backup data.

.. note::
  
  Having the prompt request for the password is more secure than including the password straight away in the command. 

> The ``--single-transaction`` `flag <https://dev.mysql.com/doc/refman/8.0/en/mysqldump.html#option_mysqldump_single-transaction>`_ starts a transaction before running. This means that the process does not lock the entire database. This allows ``mysqldump`` to read the database in the current state at the time of the transaction which ensures that your data is reliable

Reviewer: I'm adding a bit more detail here to make the sentence more understandable by the target audience (DBAs)

```suggestion
The ``--single-transaction`` `flag <https://dev.mysql.com/doc/refman/8.0/en/mysqldump.html#option_mysqldump_single-transaction>`_ starts a transaction in isolation mode ``REPEATABLE READ`` before running, allowing ``mysqldump`` to read the database in the current state at the time of the transaction and therefore ensuring consistency of the data.
```

Backup and restore MySQL data using ``mysqldump``
=================================================

> Performing the backup of your MySQL data to another storage service is a good way to ensure access to your data in case a failure occurs. In this article, you can find out how to copy your Aiven for MySQL data to a file and backup to another Aiven for MySQL database using ``mysqldump`` `tool <https://dev.mysql.com/doc/refman/8.0/en/mysqldump.html>`__.

Reviewer 2: ```suggestion
Backing up your MySQL data to another storage service is a good way to ensure access to your data in case a failure occurs. This article shows you how to copy your Aiven for MySQL data to a file, back it up to another Aiven for MySQL database, and restore it using the ``mysqldump`` `tool <https://dev.mysql.com/doc/refman/8.0/en/mysqldump.html>`__.
```

Prerequisites
-------------

* The ``mysqldump`` `tool <https://dev.mysql.com/doc/refman/8.0/en/mysqldump.html>`_ installed. Check out the `official MySQL <https://dev.mysql.com/doc/mysql-shell/8.0/en/mysql-shell-install.html>`_ documentation on how to install it.
  

> * A MySQL database that we aim to copy our data from. We will refer to it as ``source-db``.

Reviewer 2: ```suggestion
* A source MySQL database to copy your data from. We will refer to it as ``source-db``.
```

> * A target MySQL database that we aim to dump our ``source-db`` data. We will refer to it as ``target-db``.

Reviewer 2: ```suggestion
* A target MySQL database to dump your ``source-db`` data to. We will refer to it as ``target-db``.
```

.. tip::

    For the restore process, we recommend you pick a plan size that is large enough to store your data, so you can limit the downtime if you're performing a migration.

> The below example uses Aiven for MySQL databases for both, ``source-db`` and ``target-db``. You can create the databases by going to your running service for **Aiven for MySQL** > **Databases** > **Create a new database**.

Reviewer 2: ```suggestion
The below example uses Aiven for MySQL databases for both the ``source-db`` and ``target-db``. You can create the databases by going to your running service for **Aiven for MySQL** > **Databases** > **Create a new database**.
```

> Backup the data

Reviewer 2: ```suggestion
Back up the data
```

---------------

Variables
'''''''''

> To backup the ``source-db`` data to a file called ``mydb_backup.sql`` you need to collect some information about your Aiven for MySQL ``source-db`` database. Go to your Aiven ``source-db`` service, from **Overview** page, find the following information:

Reviewer 2: ```suggestion
To back up the ``source-db`` data, you need to include the connection information for your Aiven for MySQL service in the ``mysqldump`` command. Go to your Aiven ``source-db`` service, on the **Overview** page find the following information:
```

.. list-table::
   :widths: 20 80
   :header-rows: 1

   * - Variable
     - Description
   * - ``SOURCE_DB_HOST``

>      - **Host** name for the connection

Reviewer 2: Should we make these the same as they are shown in the Console? That is: Host, User, Port, Password, Database Name

   * - ``SOURCE_DB_USER``
     - **User** name for the connection
   * - ``SOURCE_DB_PORT``
     - Connection **Port** number
   * - ``SOURCE_DB_PASSWORD``
     - Connection **Password**
   * - ``DEFAULTDB``
     - Database that contains the ``source-db`` data

Commands
'''''''''

> Use the following command to backup your Aiven for MySQL data to the ``mydb_backup.sql`` file:

Reviewer 2: ```suggestion
Use the following command to back up your Aiven for MySQL data to a file named ``mydb_backup.sql``:
```

.. code-block:: shell

    mysqldump \
    -p DEFAULTDB -P SOURCE_DB_PORT \
    -h SOURCE_DB_HOST --single-transaction \
    -u SOURCE_DB_USER --set-gtid-purged=OFF \
    --password > mydb_backup.sql

> With this command, the password will be requested at the prompt; paste ``SOURCE_DB_PASSWORD`` to the terminal, then a file named ``mydb_backup.sql`` will be created with your backup data. Note that having the prompt request for the password is more secure than including the password straight away in the command. 

Reviewer 2: ```suggestion
With this command, the password will be requested at the prompt; paste ``SOURCE_DB_PASSWORD`` into the terminal, then a file named ``mydb_backup.sql`` will be created with your backup data. Note that having the prompt request for the password is more secure than including the password straight away in the command. 
```

> The ``--single-transaction`` `flag <https://dev.mysql.com/doc/refman/8.0/en/mysqldump.html#option_mysqldump_single-transaction>`_ starts a transaction in isolation mode ``REPEATABLE READ`` before running, allowing ``mysqldump`` to read the database in the current state at the time of the transaction and therefore ensuring consistency of the data.

Reviewer 2: ```suggestion
The ``--single-transaction`` `flag <https://dev.mysql.com/doc/refman/8.0/en/mysqldump.html#option_mysqldump_single-transaction>`_ starts a transaction in isolation mode ``REPEATABLE READ`` before running. This allows ``mysqldump`` to read the database in its current state at the time of the transaction, ensuring consistency of the data.
```

Reviewer 2: Just broke this up for readability since it is a little complex.

.. warning::

  If you are using `Global Transaction Identifiers <https://dev.mysql.com/doc/refman/5.7/en/replication-gtids-concepts.html>`_ (GTIDs) with InnoDB use the ``--set-gtid-purged=OFF`` `option <https://dev.mysql.com/doc/refman/8.0/en/mysqldump.html#option_mysqldump_set-gtid-purged>`_. The reason is that GTID's are not available with MyISAM.

  

Restore the data
----------------

Variables
'''''''''

> You can back up the data previously saved in a file to an Aiven for MySQL database. You need to collect some information about your Aiven for MySQL ``target-db`` to restore the data. Go to your Aiven ``target-db`` service, from **Overview** page, find the following information:

Reviewer 2: ```suggestion
To restore data saved in a file to an Aiven for MySQL database, you need to include the connection information for the service in the ``mysqldump`` command. Go to your Aiven ``target-db`` service, on the **Overview** page find the following information:
```

.. list-table::
   :widths: 20 80
   :header-rows: 1

   * - Variable
     - Description
   * - ``TARGET_DB_HOST``

>      - Service host name

Reviewer 2: Same comment as above. Either way, since it is the same information we are getting from the Console it would be good to have the naming/formatting the same in both tables.

Author: I think this is a great idea, I changed here: https://github.com/aiven/devportal/pull/1345/commits/d70215aec71ee82736fb18c6f40d5430742ee6b6

   * - ``TARGET_DB_USER``
     - Service user name
   * - ``TARGET_DB_PORT``
     - Port number to use
   * - ``TARGET_DB_PASSWORD``
     - Connection password
   * - ``DEFAULTDB``
     - Database that contains the ``target-db`` data

Commands
'''''''''

> Run the following command to store your saved data to your Aiven for MySQL database:

Reviewer 2: ```suggestion
Run the following command to load your saved data into your Aiven for MySQL database:
```

.. code-block:: shell

    mysqldump \
    -p DEFAULTDB -P TARGET_DB_PORT \
    -h TARGET_DB_HOST \
    -u TARGET_DB_USER \
    --password < mydb_backup.sql

> The password will be requested at the prompt. You can paste ``TARGET_DB_PASSWORD`` to the terminal. Your data should be stored in your Aiven for MySQL ``target-db``. You can check out ``mysqlcheck`` `command <https://dev.mysql.com/doc/refman/8.0/en/mysqlcheck.html>`_ to perform further analyzes of your current MySQL data.

Reviewer 2: ```suggestion
The password will be requested at the prompt. You can paste ``TARGET_DB_PASSWORD`` into the terminal. Your data should be stored in your Aiven for MySQL ``target-db``. You can check out ``mysqlcheck`` `command <https://dev.mysql.com/doc/refman/8.0/en/mysqlcheck.html>`_ to perform further analysis of your current MySQL data.
```
